src/onnx/parse_attention.cpp

```diff
@@ -0,0 +1,967 @@
+        }*/
+
+        return final_mask;
+    }
+
+    /* static instruction_ref handle_past_present_keys(const onnx_parser::node_info& info,
+                                                 const instruction_ref& past_input,
+                                                 const instruction_ref& keys,
+                                                 const instruction_ref& values,
+                                                 const attention_attr& parsed_attributes)
+    {
+        // Just reuse the past input if past_present are shared.
+        if(parsed_attributes.past_present_share_buffer)
+        {
+            return past_input;
+        }
+        else
+        {
+            // In a non-buffer-sharing scenario, we need to create the present state
+            // by combining keys and values and stacking them. key/values are given as
+            // (batch, sequence_length, head_size, num_heads)  head_size = querry size in some
+    literature for attention
+            // The output is expected to be a concat of shape (2, batch, num_heads, sequence_length,
+    head_size) auto value_trans = info.add_instruction(make_op("transpose", {{"permutation", {0, 3 ,
+    1, 2}}}), values); value_trans = info.add_instruction(make_op("unsqueeze", {{"axes", {0}}}),
+    value_trans);
+
+            auto keys_trans  = info.add_instruction(make_op("transpose", {{"permutation", {0, 3 , 1,
+    2}}}), keys); keys_trans  = info.add_instruction(make_op("unsqueeze", {{"axes", {0}}}),
+    keys_trans);
+
+            return info.add_instruction(make_op("concat"), keys, values);
+        }
+    } */
+
+    std::vector<instruction_ref> parse(const op_desc& /*opd*/,
+                                       const onnx_parser& parser,
+                                       const onnx_parser::node_info& info,
+                                       const std::vector<instruction_ref>& args) const
+    {
+        struct attention_attr parsed_attributes;
+        struct attention_inferred inferred_attributes;
+
+        handle_attributes(parser, info, parsed_attributes, inferred_attributes);
+        auto inputs = handle_arguments(parser, args, parsed_attributes, inferred_attributes);
+
+        auto input_data = inputs.at(0);
+        auto weights    = inputs.at(1);
+        // Set projection bias when parsed in
+        instruction_ref input_bias;
+        if(inferred_attributes.has_input_bias)
+            input_bias = inputs.at(2);
+
+        // Apply linear stage to QKV mats from weight matrix - If past input just return q mat later
+        // split will be extracted from past vector
+        auto qkv_mats = input_linear_to_qkv(info,
+                                            input_data,
+                                            weights,
+                                            parsed_attributes.qkv_hidden_sizes,
+                                            input_bias,
```

Comment:
I dont think its good to pass this as uninitialized. Instead make them std::optional<instruction_ref> and you can drop passing the additional boolean `inferred_attributes.has_input_bias`.

Reply:
Good idea and I did this. Cleans this up nicely